File: backend/src/scripts/old_importers/build_dsa_bank.py
import csv, json, re, zipfile
from pathlib import Path

ROOT = Path.cwd()
DATASETS = ROOT / "src" / "data" / "datasets"
OUT = ROOT / "src" / "data" / "dsaCodingQuestions.json"

TOPIC_ORDER = [
    "Arrays","Strings","Linked List","Stack","Queue","Trees","Binary Search Tree",
    "Graphs","Dynamic Programming","Greedy","Recursion","Backtracking","Searching",
    "Sorting","Hashing","Heap / Priority Queue","Sliding Window","Two Pointers",
    "Bit Manipulation","Math / Number Theory"
]

PLATFORM_FILES = ["leetcode", "codechef", "codeforces", "gfg", "gfg problem", "geeksforgeeks", "geeks for geeks", "problem statements"]

BAD = [
    "analogy","preposition","grammar","synonym","antonym","aptitude","profit","loss",
    "percentage","kg","grams","jwt","cors","bcrypt","rest api","microservice","react",
    "redux","hook","sap","hana","foreign currency","valuation","machine learning",
    "random forest","k means","json.dumps","json.loads","git status","cloud region"
]

def low(x): return str(x or "").lower()

def title_clean(x):
    s = str(x or "").strip()
    s = re.sub(r"^problem\s*\d+\s*:\s*", "", s, flags=re.I)
    s = re.sub(r"[_]+", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s.title()

def slugify(x):
    s = low(x)
    s = re.sub(r"[^a-z0-9\s-]", "", s)
    s = re.sub(r"\s+", "-", s)
    return re.sub(r"-+", "-", s).strip("-")

def pick(row, names):
    if not isinstance(row, dict): return ""
    keys = list(row.keys())
    for n in names:
        for k in keys:
            if low(k).strip() == low(n).strip() and row.get(k):
                return str(row.get(k))
    for n in names:
        for k in keys:
            if low(n).strip() in low(k).strip() and row.get(k):
                return str(row.get(k))
    return ""

def read_csv(text):
    try:
        return [dict(r) for r in csv.DictReader(text.splitlines())]
    except Exception:
        return []

def read_json(text):
    try:
        data = json.loads(text)
        if isinstance(data, list): return data
        if isinstance(data, dict):
            for k in ["data","questions","rows","problems","items","results"]:
                if isinstance(data.get(k), list): return data[k]
    except Exception:
        pass
    return []

def read_file(path):
    rows = []
    if path.suffix.lower() == ".zip":
        try:
            with zipfile.ZipFile(path) as z:
                for m in z.namelist():
                    ml = m.lower()
                    if not (ml.endswith(".csv") or ml.endswith(".json")):
                        continue
                    text = z.read(m).decode("utf-8", errors="ignore")
                    part = read_csv(text) if ml.endswith(".csv") else read_json(text)
                    for r in part:
                        if isinstance(r, dict):
                            r["_source_file"] = path.name
                            r["_inner_file"] = m
                            rows.append(r)
        except Exception as e:
            print("ZIP read failed:", path.name, e)
        return rows

    try:
        text = path.read_text(encoding="utf-8", errors="ignore")
        part = read_csv(text) if path.suffix.lower() == ".csv" else read_json(text)
        for r in part:
            if isinstance(r, dict):
                r["_source_file"] = path.name
                rows.append(r)
    except Exception as e:
        print("Read failed:", path.name, e)
    return rows

def platform_from_file(path, row, title):
    raw = " ".join([
        path.name, pick(row, ["platform","site","source","judge"]),
        pick(row, ["url","link","problem_link","problem url","question link"]),
        title
    ]).lower()
    if "leetcode" in raw: return "LeetCode"
    if "codechef" in raw or "code chef" in raw: return "CodeChef"
    if "codeforces" in raw or "code forces" in raw: return "Codeforces"
    if "gfg" in raw or "geeksforgeeks" in raw or "geeks for geeks" in raw: return "GeeksforGeeks"
    return "GeeksforGeeks"

def difficulty(row, i):
    raw = low(pick(row, ["difficulty","level","difficulty_level"]))
    if "easy" in raw: return "Easy"
    if "hard" in raw: return "Hard"
    if "medium" in raw: return "Medium"
    return ["Easy","Medium","Medium","Hard"][i % 4]

def topic(row, title):
    raw = " " + re.sub(r"[-_/]+", " ", " ".join([
        title,
        pick(row, ["topic","topics","tag","tags","category","subtopic","pattern"]),
        pick(row, ["url","link","problem_link"])
    ]).lower()) + " "

    checks = [
        ("Sliding Window", ["sliding window","minimum window","longest substring","window maximum"]),
        ("Two Pointers", ["two pointer","two pointers","3sum","three sum","4sum","container with most water"]),
        ("Linked List", ["linked list","listnode","list node","reverse list","merge list","cycle in linked"]),
        ("Binary Search Tree", ["binary search tree","bst"]),
        ("Trees", ["binary tree","tree","traversal","inorder","preorder","postorder","level order","diameter"]),
        ("Heap / Priority Queue", ["heap","priority queue","top k","kth largest","kth smallest"]),
        ("Dynamic Programming", ["dynamic programming"," dp ","memoization","tabulation","knapsack","coin change","lis","lcs","edit distance","house robber","climbing stairs"]),
        ("Graphs", ["graph","bfs","dfs","dijkstra","bellman","floyd","topological","island","course schedule","union find"]),
        ("Backtracking", ["backtracking","n queen","n-queen","sudoku","rat in a maze","word search","combination sum","permutation","subsets"]),
        ("Recursion", ["recursion","recursive","tower of hanoi"]),
        ("Greedy", ["greedy","activity selection","job sequencing","jump game","gas station","candy"]),
        ("Searching", ["binary search","search insert","rotated sorted","peak element","first and last"]),
        ("Sorting", ["sort","sorting","merge sort","quick sort","count inversions","merge intervals"]),
        ("Hashing", ["hash","hashmap","hash map","duplicate","anagram","two sum"]),
        ("Stack", ["stack","parentheses","bracket","next greater","histogram","min stack"]),
        ("Queue", ["queue","deque","circular queue","rotten oranges"]),
        ("Strings", ["string","substring","palindrome","anagram","prefix","suffix","roman"]),
        ("Bit Manipulation", ["bit","xor","single number","power of two"]),
        ("Math / Number Theory", ["math","prime","gcd","lcm","sieve","factorial"]),
        ("Arrays", ["array","subarray","matrix","two sum","maximum subarray","rotate array","missing number"])
    ]
    for t, words in checks:
        if any(w in raw for w in words):
            return t
    return "Arrays"

def direct_url(row, title, plat):
    direct = pick(row, ["url","URL","link","Link","problem_link","problem_url","Problem Link","question link","Question Link","Problem URL"])
    if direct.startswith("http"):
        return direct

    code = pick(row, ["code","problem_code","problem code","contest_code","slug","titleSlug","title_slug","problem_slug"]).strip()
    contest = pick(row, ["contestId","contest_id","contest"]).strip()
    idx = pick(row, ["index","problemIndex","problem_index"]).strip()
    s = slugify(code or title)

    if plat == "LeetCode":
        return f"https://leetcode.com/problems/{s}/"
    if plat == "CodeChef":
        if code and " " not in code and len(code) <= 40:
            return f"https://www.codechef.com/problems/{code.upper()}"
        return f"https://www.codechef.com/practice?search={slugify(title)}"
    if plat == "Codeforces":
        if contest and idx:
            return f"https://codeforces.com/problemset/problem/{contest}/{idx}"
        return f"https://codeforces.com/problemset?search={title.replace(' ', '+')}"
    if plat == "GeeksforGeeks":
        return f"https://www.geeksforgeeks.org/search/{title.replace(' ', '%20')}/"
    return f"https://www.google.com/search?q={title.replace(' ', '+')}+DSA+problem"

def acceptable_title(title, row, source):
    raw = low(title + " " + json.dumps(row, ensure_ascii=False) + " " + source)
    if any(b in raw for b in BAD):
        return False
    if len(title) < 3:
        return False
    # Platform datasets are already coding datasets; keep broad but remove obvious theory prompts.
    if low(title).startswith(("what is ", "what are ", "explain ", "describe ", "discuss ", "define ", "choose ")):
        known = ["two sum","valid parentheses","coin change","merge intervals","number of islands"]
        return any(k in raw for k in known)
    return True

files = []
for base in [DATASETS, ROOT/"src"/"data", ROOT/"data"]:
    if base.exists():
        for p in base.rglob("*"):
            if p.is_file() and p.suffix.lower() in [".zip",".csv",".json"]:
                n = p.name.lower()
                if any(x in n for x in PLATFORM_FILES):
                    files.append(p)

files = sorted(set(files))
seen = set()
out = []

print("Used files:")
for f in files:
    print("-", f.name)

for file in files:
    rows = read_file(file)
    print(f"Rows from {file.name}: {len(rows)}")

    for i, row in enumerate(rows):
        raw_title = pick(row, [
            "title","Title","problem","problem_name","problem name","Problem Name",
            "problem_title","Problem Title","question_title","question","Question",
            "name","Name","titleSlug","title_slug","slug","problem_name","Problem Name","problem_title","Problem Title","problemName","problemTitle"
        ])

        title = title_clean(raw_title)
        if not acceptable_title(title, row, file.name):
            continue

        plat = platform_from_file(file, row, title)
        top = topic(row, title)
        diff = difficulty(row, i)
        url = direct_url(row, title, plat)

        key = f"{title}|{plat}|{top}".lower()
        if key in seen:
            continue
        seen.add(key)

        out.append({
            "id": f"dsa-{len(out)+1}",
            "title": title,
            "topic": top,
            "platform": plat,
            "difficulty": diff,
            "url": url,
            "sourceDataset": file.name
        })

topic_index = {t:i for i,t in enumerate(TOPIC_ORDER)}
diff_index = {"Easy":1,"Medium":2,"Hard":3}

out.sort(key=lambda x: (
    topic_index.get(x["topic"], 999),
    x["platform"],
    diff_index.get(x["difficulty"], 2),
    x["title"]
))

OUT.parent.mkdir(parents=True, exist_ok=True)
OUT.write_text(json.dumps(out, indent=2), encoding="utf-8")

print(f"DSA coding bank created: {len(out)} questions")
print(f"Saved to: {OUT}")

File: backend/src/scripts/buildDsaBank.js
const fs = require("fs");
const path = require("path");

const DATA_DIRS = [
  path.join(process.cwd(), "src", "data", "datasets"),
  path.join(process.cwd(), "src", "data"),
  path.join(process.cwd(), "data")
];

const OUT = path.join(process.cwd(), "src", "data", "dsaCodingQuestions.json");

const TOPIC_ORDER = [
  "Arrays",
  "Strings",
  "Linked List",
  "Stack",
  "Queue",
  "Trees",
  "Binary Search Tree",
  "Graphs",
  "Dynamic Programming",
  "Greedy",
  "Recursion",
  "Backtracking",
  "Searching",
  "Sorting",
  "Hashing",
  "Heap / Priority Queue",
  "Sliding Window",
  "Two Pointers",
  "Bit Manipulation",
  "Math / Number Theory"
];

function low(x) {
  return String(x || "").toLowerCase();
}

function cleanTitle(s) {
  return String(s || "")
    .replace(/[_]+/g, " ")
    .replace(/\s+/g, " ")
    .replace(/^problem\s*\d+\s*:\s*/i, "")
    .trim()
    .replace(/\b\w/g, (m) => m.toUpperCase());
}

function walk(dir, files = []) {
  if (!fs.existsSync(dir)) return files;

  for (const item of fs.readdirSync(dir)) {
    const full = path.join(dir, item);
    const stat = fs.statSync(full);

    if (stat.isDirectory()) walk(full, files);
    else if (/\.(csv|json)$/i.test(full)) files.push(full);
  }

  return files;
}

function parseCSV(text) {
  const rows = [];
  let row = [];
  let cell = "";
  let quote = false;

  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    const n = text[i + 1];

    if (c === '"' && quote && n === '"') {
      cell += '"';
      i++;
    } else if (c === '"') {
      quote = !quote;
    } else if (c === "," && !quote) {
      row.push(cell);
      cell = "";
    } else if ((c === "\n" || c === "\r") && !quote) {
      if (cell || row.length) {
        row.push(cell);
        rows.push(row);
      }
      row = [];
      cell = "";
      if (c === "\r" && n === "\n") i++;
    } else {
      cell += c;
    }
  }

  if (cell || row.length) {
    row.push(cell);
    rows.push(row);
  }

  if (!rows.length) return [];

  const headers = rows[0].map((h) => String(h || "").trim());
  return rows.slice(1).map((r) => {
    const obj = {};
    headers.forEach((h, i) => {
      obj[h] = r[i] || "";
    });
    return obj;
  });
}

function loadFile(file) {
  try {
    const text = fs.readFileSync(file, "utf8");

    if (/\.json$/i.test(file)) {
      const data = JSON.parse(text);
      if (Array.isArray(data)) return data;
      if (Array.isArray(data.data)) return data.data;
      if (Array.isArray(data.questions)) return data.questions;
      if (Array.isArray(data.problems)) return data.problems;
      if (Array.isArray(data.rows)) return data.rows;
      return [];
    }

    return parseCSV(text);
  } catch {
    return [];
  }
}

function getAny(row, keys) {
  const all = Object.keys(row || {});

  for (const key of keys) {
    const exact = all.find((k) => low(k) === low(key));
    if (exact && row[exact]) return row[exact];
  }

  for (const key of keys) {
    const partial = all.find((k) => low(k).includes(low(key)));
    if (partial && row[partial]) return row[partial];
  }

  return "";
}

function useOnlyProgrammingCodingDataset(file) {
  const f = low(file);

  const reject = [
    "computer science theory",
    "hr interview",
    "interview questions hr",
    "indian job market",
    "prepmaster placement",
    "placement",
    "qa dataset",
    "theory",
    "aptitude",
    "mcq"
  ];

  if (reject.some((x) => f.includes(x))) return false;

  const accept = [
    "dataset of programming questions",
    "programming questions",
    "coding",
    "dsa",
    "leetcode",
    "gfg",
    "geeksforgeeks",
    "hackerrank",
    "interviewbit",
    "software"
  ];

  return accept.some((x) => f.includes(x));
}

function platformFrom(row, file, title, index) {
  const s = low([
    getAny(row, ["platform", "site", "source", "judge", "company"]),
    getAny(row, ["url", "link", "problem_link", "problem url"]),
    file,
    title
  ].join(" "));

  if (s.includes("leetcode")) return "LeetCode";
  if (s.includes("geeksforgeeks") || s.includes("gfg")) return "GeeksforGeeks";
  if (s.includes("hackerrank")) return "HackerRank";
  if (s.includes("interviewbit")) return "InterviewBit";

  return index % 2 === 0 ? "LeetCode" : "GeeksforGeeks";
}

function difficultyFrom(row, title, index) {
  const s = low([getAny(row, ["difficulty", "level"]), title].join(" "));

  if (s.includes("hard")) return "Hard";
  if (s.includes("easy")) return "Easy";
  if (s.includes("medium")) return "Medium";

  if (index % 10 < 4) return "Easy";
  if (index % 10 < 8) return "Medium";
  return "Hard";
}

function topicFrom(row, title) {
  const s = low([
    getAny(row, ["topic", "topics", "tag", "tags", "category", "subtopic"]),
    title
  ].join(" "));

  if (s.includes("sliding window")) return "Sliding Window";
  if (s.includes("two pointer") || s.includes("2 pointer") || s.includes("3sum") || s.includes("container with most water")) return "Two Pointers";
  if (s.includes("linked list")) return "Linked List";
  if (s.includes("binary search tree") || s.includes("bst")) return "Binary Search Tree";
  if (s.includes("binary tree") || s.includes("tree") || s.includes("traversal")) return "Trees";
  if (s.includes("priority queue") || s.includes("heap")) return "Heap / Priority Queue";
  if (s.includes("dynamic programming") || s.includes(" dp") || s.includes("knapsack") || s.includes("coin change") || s.includes("lcs") || s.includes("lis")) return "Dynamic Programming";
  if (s.includes("graph") || s.includes("bfs") || s.includes("dfs") || s.includes("island") || s.includes("dijkstra") || s.includes("topological")) return "Graphs";
  if (s.includes("backtracking") || s.includes("n queen") || s.includes("n-queen") || s.includes("sudoku") || s.includes("rat in a maze")) return "Backtracking";
  if (s.includes("recursion") || s.includes("recursive") || s.includes("subsets") || s.includes("permutation")) return "Recursion";
  if (s.includes("greedy") || s.includes("activity selection") || s.includes("job sequencing") || s.includes("jump game")) return "Greedy";
  if (s.includes("binary search") || s.includes("search")) return "Searching";
  if (s.includes("sort") || s.includes("merge intervals") || s.includes("inversion")) return "Sorting";
  if (s.includes("hash") || s.includes("map") || s.includes("duplicate") || s.includes("anagram")) return "Hashing";
  if (s.includes("stack") || s.includes("parentheses") || s.includes("next greater") || s.includes("histogram")) return "Stack";
  if (s.includes("queue") || s.includes("deque")) return "Queue";
  if (s.includes("string") || s.includes("palindrome") || s.includes("substring") || s.includes("anagram")) return "Strings";
  if (s.includes("bit") || s.includes("xor")) return "Bit Manipulation";
  if (s.includes("math") || s.includes("number theory") || s.includes("prime") || s.includes("gcd") || s.includes("lcm")) return "Math / Number Theory";
  if (s.includes("array") || s.includes("subarray") || s.includes("matrix") || s.includes("two sum")) return "Arrays";

  return "Arrays";
}

function isBadNonCoding(title, row, file) {
  const s = low(title + " " + JSON.stringify(row) + " " + file);

  const bad = [
    "complete the analogy",
    "analogy",
    "preposition",
    "grammar",
    "synonym",
    "antonym",
    "aptitude",
    "profit",
    "loss",
    "percentage",
    "probability",
    "coin is tossed",
    "committee",
    "people sit",
    "sit in a row",
    "boxes are stacked",
    "fish :",
    "egg:",
    "json.dumps",
    "json.loads",
    "os.path",
    "numpy",
    "git status",
    "git diff",
    "git checkout",
    "cloud region",
    "google cloud",
    "new relic",
    "tls",
    "cdn",
    "random forest",
    "k means",
    "knn algorithm",
    "decision stump",
    "machine learning",
    "classifier",
    "regression",
    "backend",
    "jwt",
    "cors",
    "bcrypt",
    "restful api",
    "microservice",
    "web server",
    "authentication",
    "authorization",
    "react",
    "redux",
    "hook",
    "sap",
    "hana",
    "sales",
    "billing",
    "valuation",
    "foreign currency",
    "database challenge",
    "oop concepts challenge",
    "solid principles",
    "blockchain",
    "compiler for a new programming language",
    "quantum algorithm",
    "ai that can play",
    "messaging system"
  ];

  return bad.some((x) => s.includes(x));
}

function isCodingProblem(title, row, file) {
  const s = low(title + " " + JSON.stringify(row) + " " + file);

  if (isBadNonCoding(title, row, file)) return false;

  const codingSignals = [
    "leetcode",
    "geeksforgeeks",
    "gfg",
    "hackerrank",
    "interviewbit",
    "codechef",
    "codeforces",
    "array",
    "string",
    "linked list",
    "stack",
    "queue",
    "tree",
    "graph",
    "heap",
    "greedy",
    "recursion",
    "backtracking",
    "dynamic programming",
    "binary search",
    "sorting",
    "searching",
    "hashing",
    "sliding window",
    "two pointer",
    "bit manipulation",
    "two sum",
    "maximum subarray",
    "minimum jumps",
    "minimize the heights",
    "count inversions",
    "kth largest",
    "merge intervals",
    "valid parentheses",
    "next greater",
    "largest rectangle",
    "rotate array",
    "missing number",
    "majority element",
    "product of array",
    "palindrome",
    "anagram",
    "substring",
    "reverse linked",
    "detect loop",
    "merge two sorted",
    "remove nth",
    "bfs",
    "dfs",
    "dijkstra",
    "number of islands",
    "coin change",
    "knapsack",
    "longest common subsequence",
    "longest increasing subsequence",
    "climbing stairs",
    "house robber",
    "edit distance",
    "jump game",
    "activity selection",
    "job sequencing",
    "n-queens",
    "sudoku",
    "rat in a maze",
    "subsets",
    "permutations",
    "combination sum",
    "quick sort",
    "merge sort",
    "trie",
    "xor"
  ];

  return codingSignals.some((x) => s.includes(x));
}

function slugify(title) {
  return low(title)
    .replace(/[^a-z0-9\s-]/g, "")
    .replace(/\s+/g, "-")
    .replace(/-+/g, "-")
    .replace(/^-|-$/g, "");
}

function directUrl(row, title, platform) {
  const direct = getAny(row, ["url", "link", "problem_link", "problem url", "question link"]);
  if (direct && /^https?:\/\//i.test(direct)) return direct;

  const slug = slugify(title);

  if (platform === "LeetCode") {
    return `https://leetcode.com/problems/${slug}/`;
  }

  if (platform === "GeeksforGeeks") {
    return `https://www.geeksforgeeks.org/problems/${slug}/1`;
  }

  if (platform === "HackerRank") {
    return `https://www.hackerrank.com/search?term=${encodeURIComponent(title)}`;
  }

  if (platform === "InterviewBit") {
    return `https://www.interviewbit.com/search/?q=${encodeURIComponent(title)}`;
  }

  return `https://leetcode.com/problems/${slug}/`;
}

const files = Array.from(new Set(DATA_DIRS.flatMap((d) => walk(d)))).filter(useOnlyProgrammingCodingDataset);

const seen = new Set();
const out = [];

for (const file of files) {
  const rows = loadFile(file);
  const fileName = path.basename(file);

  rows.forEach((row, index) => {
    const titleRaw = getAny(row, [
      "title",
      "problem",
      "problem_name",
      "problem name",
      "question_title",
      "question",
      "name"
    ]);

    const title = cleanTitle(titleRaw);
    if (!title || title.length < 3) return;
    if (!isCodingProblem(title, row, fileName)) return;

    const topic = topicFrom(row, title);
    const platform = platformFrom(row, fileName, title, index);
    const difficulty = difficultyFrom(row, title, index);

    const key = `${title}-${topic}-${platform}`.toLowerCase();
    if (seen.has(key)) return;
    seen.add(key);

    out.push({
      id: `dsa-${out.length + 1}`,
      title,
      topic,
      platform,
      difficulty,
      url: directUrl(row, title, platform),
      sourceDataset: fileName
    });
  });
}

out.sort((a, b) => {
  const ta = TOPIC_ORDER.indexOf(a.topic);
  const tb = TOPIC_ORDER.indexOf(b.topic);

  return (
    (ta === -1 ? 999 : ta) - (tb === -1 ? 999 : tb) ||
    a.platform.localeCompare(b.platform) ||
    a.difficulty.localeCompare(b.difficulty) ||
    a.title.localeCompare(b.title)
  );
});

fs.writeFileSync(OUT, JSON.stringify(out, null, 2));

console.log("Used dataset files:");
files.forEach((f) => console.log("-", path.basename(f)));
console.log(`DSA coding bank created: ${out.length} questions`);
console.log(`Saved to: ${OUT}`);
